text/1615-cargo-platform-standards.md

```diff
@@ -0,0 +1,170 @@
+  add its application-private clutter to it. _(Linux.)_
+* Using a standard directory for binary outputs can allow the user to execute
+  Cargo-installed binaries without modifying their `PATH` variable. _(Linux)_
+
+Solving this problem will likely also solve the same problem in Cargo-related
+tools such as `rustup` as their strategy is "do what Cargo does".
+
+This seems to be implemented in pip, the Python package manager, already.
+
+# Detailed design
+
+We are going to introduce new environment variables:
+```
+CARGO_BIN_DIR
+CARGO_CACHE_DIR
+CARGO_CONFIG_DIR
+```
+
+For the default values of these variables if they are set to the empty string
+or not set (which will be the common case), see below.
+
+These will be used to split the current `.cargo` (`CARGO_HOME`) directory up:
+The cached packages (`.cargo/git`, `.cargo/registry`) will go into
+`CARGO_CACHE_DIR`, binaries (`.cargo/bin`) installed by Cargo will go into
+`CARGO_BIN_DIR` and the config (`.cargo/config`) will go into
+`CARGO_CONFIG_DIR`.
+
+In order to maintain backward compatibility, the old directory locations will
+be checked if the new ones don't exist. In detail, this means:
+
+1. If any of the new variables `CARGO_BIN_DIR`, `CARGO_CACHE_DIR`,
+   `CARGO_CONFIG_DIR` are set and nonempty, use the new directory structure.
+2. Else, if there is an override for the legacy Cargo directory, using
+   `CARGO_HOME`, the directories for cache, configuration and executables are
+   placed inside this directory.
+3. Otherwise, if the Cargo-specfic platform-specific directories exist, use
+   them. What constitutes a Cargo-specific directory is laid out below, for
+   each platform.
+4. If that's not the case, check whether the legacy directory exists (`.cargo`)
+   and use it in that case.
+5. If everything else fails, create the platform-specific directories and use
+   them.
+
+This makes Cargo use platform-specific directories for new installs while
+retaining compatibility for the old directory layout. It also allows one to
+keep all Cargo related data in one place if one wishes to.
+
+## Windows
+
+We'll obtain each of the following directories using the correct API.
+
+```
+cache:    AppData\Local\Temp\Cargo
+config:   AppData\Roaming\Cargo
+binaries: AppData\Local\Programs\Cargo
+```
+
+## Unixy systems (macOS, Linux, BSDs)
+
+Here, we're following the [XDG specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.7.html).
+By default, if no further variables are set, this means that we'll be using the
+following subdirectories of the home directory.
+
+```
+cache:    .cache/cargo
+config    .config/cargo
+binaries: .local/bin
+```
+
+**There is currently an on-going discussion about standardizing the location of
+`.local/bin` together with a new XDG variable `XDG_BIN_HOME`. The
+implementation of this RFC should be delayed until that discussion has finished
+and use the result. This RFC will be amended with that result.**
```

Comment:
Has there been any news about this?

Reply:
I'd just go forward with it, because there are no negative consequences from not doing it.

The non-existence of `$XDG_BIN_HOME`, just like the possible non-existence of other XDG env vars, is not a problem.

If the XDG spec update ships unchanged, the cargo code will "just" work as people start using the env var, and if there are any changes (like a different name of the env var) the code can just be updated.

Reply:
I'm still waiting for poettering to reply to my emails.

Reply:
I'd suggest sending a follow-up, and in that follow-up I'd suggest asking if it'd be possible to move forward given the absence of objections.

Reply:
The problem is mostly just that I was depending on pottering to actually *push* it, since no one else who could showed up.  But I'm planning on poking the list about it again, yeah.